### ISSUE TITLE
Cannot move shared folder into an external storage

### ISSUE THREAD

**Issue by jnweiger** [owncloud/core#39550]
Seen with server 10.9.0 beta (and encryption 1.5.1 enabled - probably unrelated, but that is what I actually tried to break)

* user admin shares his Photos folder with group users.
* user03 (member of users) renames the received folder to shared_Photos
* user03 also has an external storage mounted as SFTP
* via web browser user03 attempts to move shared_Photos into SFTP, an error occurs:

![image](https://user-images.githubusercontent.com/1108546/144660960-902ab835-b8fe-4b65-ae92-79e07f178a37.png)

From the error message, the user has no hint why that would fail. The owncloud.log has

```
{"reqId":"Yapt0naju748l7qqajR0LgAAAAE","level":0,"time":"2021-12-03T19:19:46+00:00","remoteAddr":"2.247.252.175","user":"user03","app":"OC\\User\\Session::validateToken","method":"GET","url":"\/ocs\/v2.php\/apps\/notifications\/api\/v1\/notifications?format=json","message":"token 395c45b64f0d8c6adbd53f2338ebdaa59a161e6e9a2420f6870b69dee4e91b74815639128936d5c4161ee2c8125a5a57d296af44b4758c7ca98b02cde6e78c84 with token id 18 found, validating"}
{"reqId":"Yapt0XnrzqIPaGnI7eZxcwAAAAs","level":3,"time":"2021-12-03T19:19:46+00:00","remoteAddr":"2.247.252.175","user":"user03","app":"PHP","method":"MOVE","url":"\/remote.php\/dav\/files\/user03\/shared_Photos","message":"Trying to access array offset on value of type bool at \/var\/www\/owncloud\/apps\/files_external\/lib\/Lib\/Storage\/SFTP.php#322"}
{"reqId":"Yapt0XnrzqIPaGnI7eZxcwAAAAs","level":3,"time":"2021-12-03T19:19:46+00:00","remoteAddr":"2.247.252.175","user":"user03","app":"PHP","method":"MOVE","url":"\/remote.php\/dav\/files\/user03\/shared_Photos","message":"Trying to access array offset on value of type bool at \/var\/www\/owncloud\/apps\/files_external\/lib\/Lib\/Storage\/SFTP.php#328"}
{"reqId":"Yapt0XnrzqIPaGnI7eZxcwAAAAs","level":0,"time":"2021-12-03T19:19:46+00:00","remoteAddr":"2.247.252.175","user":"user03","app":"files","method":"MOVE","url":"\/remote.php\/dav\/files\/user03\/shared_Photos","message":"It is not allowed to move one mount point into another one"}
{"reqId":"Yapt0XnrzqIPaGnI7eZxcwAAAAs","level":0,"time":"2021-12-03T19:19:46+00:00","remoteAddr":"2.247.252.175","user":"user03","app":"webdav","method":"MOVE","url":"\/remote.php\/dav\/files\/user03\/shared_Photos","message":"Exception: HTTP\/1.1 403 There was an error while renaming the file or directory: {\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\Forbidden\",\"Message\":\"There was an error while renaming the file or directory\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(160): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->moveInto()\\n#1 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(612): Sabre\\\\DAV\\\\Tree->move()\\n#2 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/WildcardEmitterTrait.php(89): Sabre\\\\DAV\\\\CorePlugin->httpMove()\\n#3 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(472): Sabre\\\\DAV\\\\Server->emit()\\n#4 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#5 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(345): Sabre\\\\DAV\\\\Server->start()\\n#6 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#7 \\\/var\\\/www\\\/owncloud\\\/remote.php(165): require_once('\\\/var\\\/www\\\/ownclo...')\\n#8 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":492}"}
{"reqId":"612cf113-f5bb-407a-bf54-d85e1f4eb1a6","level":0,"time":"2021-12-03T19:19:48+00:00","remoteAddr":"2.247.252.175","user":"user01","app":"OC\\Authentication\\Token\\DefaultTokenProvider::updateToken","method":"PROPFIND","url":"\/remote.php\/dav\/files\/user01\/","message":"updating token 25, last check is now 1638559002"}
{"reqId":"612cf113-f5bb-407a-bf54-d85e1f4eb1a6","level":0,"time":"2021-12-03T19:19:48+00:00","remoteAddr":"2.247.252.175","user":"user01","app":"OC\\Authentication\\Token\\DefaultTokenProvider::getToken","method":"PROPFIND","url":"\/remote.php\/dav\/files\/user01\/","message":"token e2da9af034125bca799414cef76a212c3cc04a84e4e9904d02249d246e882caa773461f037beab59d5d75f89bcf77b3a4b70f16b76e7c53ea9281be2f3fec52c does not exist"}
{"reqId":"74a00133-f305-4661-823b-bf39760685ea","level":0,"time":"2021-12-03T19:19:49+00:00","remoteAddr":"2.247.252.175","user":"user04","app":"OC\\Authentication\\Token\\DefaultTokenProvider::updateToken","method":"PROPFIND","url":"\/remote.php\/dav\/files\/user04\/","message":"updating token 24, last check is now 1638559001"}
```

Expected behaviour: one of
* The move should succeed, effectively making a private copy in the other storage. (That is what the user most likely intended, - compare standard unix semantics, when moving a hardlinked object into a different device)
* the move should fail, with the correct explanation "It is not allowed to move one mount point into another one"

**Comment by jnweiger** [comment]
The same behaves more weird, when the share owner moves the shared Photos folder into a WND mount.
![image](https://user-images.githubusercontent.com/1108546/144662370-d6876f54-b075-4b53-9d19-772120e10748.png)

From this error message, one would expect that nothing was moved.

Actually, the folder was moved, and the share relation was broken. Effectively unsharing the folder with everybody.
Users still see it in their shared with you listing:

![image](https://user-images.githubusercontent.com/1108546/144662679-4eda8ea6-41ed-457a-931d-fc8ab7860982.png)

When the user clicks on the Folder 'Photos (2)' he sees his own home.
![image](https://user-images.githubusercontent.com/1108546/144663331-d425f83a-f77b-4357-9a86-80d6f4d0e7f5.png)

**Comment by phil-davis** [comment]
PR #39772 added test scenarios that demonstrate the current behaviour in master. Search for acceptance test tag `@issue-39550` 

There are some scenarios that give a 500 error. That definitely needs to be fixed. Probably it should give a 403.

Other scenarios give 403 status but the non-specific message "There was an error while renaming the file or directory". The message could be improved to be more specific about what the reason is for the error.

**Comment by phil-davis** [comment]
PR #39584 has been rebased on top of the new test scenarios and the scenarios adjusted. 

See my comment https://github.com/owncloud/core/pull/39584#issuecomment-1039041597 for the next steps.

**Comment by jnweiger** [comment]
Confirmed fixed in 10.10.0 RC2

* Error message when share reciever tries move into SFTP mount: "Could not move "Documents": It is not allowed to move one mount point into another one" OK

But:
* new Web UI
   - Error message when share owner tries move into SFTP mount: "Failed to move Documents".  BAD
  - the folder is actually moved and it is still shared.

* classic  UI:
   - Error message when share owner tries move into SFTP mount: "Could not move "Documents": You are not allowed to share /admin/files/SFTP/Documents" BAD.
  - the folder is actually moved and it is still shared!
